viewing_party/party.py

```diff
@@ -1,23 +1,139 @@
+        genre_list.append(movie["genre"])
+
+    if user_data["watched"] == []:
+        return None
+    else:
+        counter = 0
+        for genre in genre_list:
+            frequency = genre_list.count(genre)
+            if frequency > counter:
+                counter = frequency
+                most_watched_genre = genre
+            return most_watched_genre
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    unique_watched = []
+    friends_movies = set()
+    users_movies = set()
+    # We want to access "friends"
+    # and for every friend in there, we want to 
+    # access a movie in their "watched" list of
+    # movie dictionaries.
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    for movie in user_data["watched"]:
+        users_movies.add(movie["title"])
+    unique_movies = users_movies - friends_movies
+    for movie in user_data["watched"]:
+        if movie["title"] in unique_movies:
+            unique_watched.append(movie)
+    return unique_watched
+
+def get_friends_unique_watched(user_data):
+    unique_watched = []
+    friends_movies = set()
+    users_movies = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    for movie in user_data["watched"]:
+        users_movies.add(movie["title"])
+    unique_movies = friends_movies - users_movies
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in unique_movies:
+                if movie not in unique_watched:
+                    unique_watched.append(movie)
+    return unique_watched 
+    
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recommended_movies = []
+    host_list = user_data["subscriptions"]
+    unique_watched = get_friends_unique_watched(user_data)
+    for movie in unique_watched:
+        if movie["host"] in host_list:
+            recommended_movies.append(movie)
+    return recommended_movies
+
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    recs_by_genre = []
+    fav_genre = get_most_watched_genre(user_data)
+    recommended = get_friends_unique_watched(user_data)
+    for movie in recommended:
+        if movie["genre"] == fav_genre:
+            recs_by_genre.append(movie)
+    return recs_by_genre
+
+def get_rec_from_favorites(user_data):
+    unique_recs = get_unique_watched(user_data)
+    recs_from_favs = []
+    
+    for movie in unique_recs:
+        if movie in user_data["favorites"]:

```

Comment:
Like above, we could build a set of the titles in the user's favorites data to gain the benefit of those constant time `in` lookups!